plugin/session_buffer.py

```diff
@@ -414,6 +441,50 @@ def update_document_link(self, new_link: DocumentLink) -> None:
                 self.document_links.append(new_link)
                 break
 
+    # --- textDocument/diagnostic --------------------------------------------------------------------------------------
+
+    def do_document_diagnostic_async(self, view: sublime.View, version: Optional[int] = None) -> None:
+        mgr = self.session.manager()
+        if not mgr:
+            return
+        if mgr.should_ignore_diagnostics(self.last_known_uri, self.session.config):
+            return
+        if version is None:
+            version = view.change_count()
+        if self.has_capability("diagnosticProvider"):
+            params = {'textDocument': text_document_identifier(view)}  # type: DocumentDiagnosticParams
+            identifier = self.get_capability("diagnosticProvider.identifier")
+            if identifier:
+                params['identifier'] = identifier
+            if self.document_diagnostic_result_id:
+                params['previousResultId'] = self.document_diagnostic_result_id
+            self.session.send_request_async(
+                Request.documentDiagnostic(params, view),
+                self._if_view_unchanged(self.on_document_diagnostic, version),
+                self._if_view_unchanged(self._on_document_diagnostic_error, version)
+            )
+
+    def on_document_diagnostic(self, view: Optional[sublime.View], response: DocumentDiagnosticReport) -> None:
+        self.document_diagnostic_result_id = response.get('resultId')
+        if is_full_document_diagnostic_report(response):
+            self.session.m_textDocument_publishDiagnostics(
+                {'uri': self.last_known_uri, 'diagnostics': response['items']})
+        for uri, diagnostic_report in response.get('relatedDocuments', {}):
+            sb = self.session.get_session_buffer_for_uri_async(uri)
+            if sb:
+                cast(SessionBuffer, sb).on_document_diagnostic(None, cast(DocumentDiagnosticReport, diagnostic_report))
+
+    def _on_document_diagnostic_error(self, view: sublime.View, error: Error) -> None:
+        if error.code == LSPErrorCodes.ServerCancelled and is_diagnostic_server_cancellation_data(error.data) and \
+                error.data['retriggerRequest']:
+            # Retrigger the request after a short delay, but only if there were no additional changes to the buffer (in
+            # that case the request will be retriggered automatically anyway)
+            version = view.change_count()
+            sublime.set_timeout_async(self._if_view_unchanged(self.do_document_diagnostic_async, version), 500)
```

Comment:
I've updated it to look more readable by replacing `Callable` that can't express optional arguments with a custom type.

Reply:
But now `version` is not passed to `self.do_document_diagnostic_async`, no? That's why I appended `(version)` and need a lamda here (I think).

Reply:
Actually I see that I made `version` optional. So it's not needed and this should be okay.

Reply:
Should we just remove the version argument? It looks like we are not really consistent and sometimes passing it and sometimes not. And it is able to infer it automatically itself.

Reply:
I see your argument but I think we can do that extra api request in this case for the sake of clarity in our code.

Reply:
I think it's there to save some duplicate API calls, but I only added this advantage with `version = view.change_count()` in `_check_did_open` in this PR. But yeah, it's only used inconsistently by a few methods.

Reply:
To be clear, what I've meant is to remove the `version` argument from *all* functions then. I've re-added it for now, and mabye it looks ugly, but that's because of the way how `_if_view_unchanged` is defined to return a callable and accept arguments. If it should not be called with arguments, in order to make the code look nicer, then I think in the first place the `_if_view_unchanged` function should be adjusted to prevent that.